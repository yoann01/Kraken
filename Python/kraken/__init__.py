"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
