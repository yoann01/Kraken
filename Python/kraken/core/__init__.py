"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
