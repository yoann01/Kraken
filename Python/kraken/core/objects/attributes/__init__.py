"""Kraken Attributes."""

__all__ = ['attribute_group',
           'attribute',
           'bool_attribute',
           'color_attribute',
           'scalar_attribute',
           'integer_attribute',
           'string_attribute']
