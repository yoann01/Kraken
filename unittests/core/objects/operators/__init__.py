
import unittest

import test_operator
import test_kl_operator
import test_canvas_operator

loadOperatorSuite = test_operator.suite()
loadKLOperatorSuite = test_kl_operator.suite()
loadCanvasOperatorSuite = test_canvas_operator.suite()


def suite():
    suites = [
        loadOperatorSuite,
        loadKLOperatorSuite,
        loadCanvasOperatorSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
