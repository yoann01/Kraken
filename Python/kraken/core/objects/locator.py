"""Kraken - objects.locator module.

Classes:
Locator -- Locator representation.

"""

from kraken.core.objects.object_3d import Object3D


class Locator(Object3D):
    """Locator object."""

    def __init__(self, name, parent=None, flags=None, metaData=None):
        super(Locator, self).__init__(name, parent=parent, flags=flags, metaData=metaData)
