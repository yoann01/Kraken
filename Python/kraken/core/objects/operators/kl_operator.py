"""Kraken - objects.operators.kl_operator module.

Classes:
KLOperator - Splice operator object.

"""

import pprint
import re

from kraken.core.maths import MathObject, Mat44, Xfo, Vec2, Vec3
from kraken.core.objects.object_3d import Object3D
from kraken.core.objects.operators.operator import Operator
from kraken.core.objects.attributes.attribute import Attribute
from kraken.core.kraken_system import ks
from kraken.log import getLogger

logger = getLogger('kraken')


class KLOperator(Operator):
    """KL Operator representation."""

    def __init__(self, name, solverTypeName, extension, metaData=None):
        super(KLOperator, self).__init__(name, metaData=metaData)

        self.solverTypeName = solverTypeName
        self.extension = extension

        # Load the Fabric Engine client and construct the RTVal for the Solver
        ks.loadCoreClient()
        ks.loadExtension('Kraken')
        if self.extension != 'Kraken':
            ks.loadExtension(self.extension)
        self.solverRTVal = ks.constructRTVal('%s::%s' % (self.extension, self.solverTypeName))

        # logger.debug("Creating kl operator object [%s] of type [%s] from extension [%s]:" % (self.getName(), self.solverTypeName, self.extension))

        self.args = self.solverRTVal.getArguments('Kraken::KrakenSolverArg[]')

        # Initialize the inputs and outputs based on the given args.
        for i in xrange(len(self.args)):
            arg = self.args[i]
            argName = arg.name.getSimpleType()
            argDataType = arg.dataType.getSimpleType()
            argConnectionType = arg.connectionType.getSimpleType()

            # Note, do not create empty arrays here as we need to know later whether or not
            # to create default values if input/output is None
            if argConnectionType == 'In':
                self.inputs[argName] = None
            else:
                self.outputs[argName] = None


    def getSolverTypeName(self):
        """Returns the solver type name for this operator.

        Returns:
            str: Name of the solver type this operator uses.

        """

        return self.solverTypeName

    def getExtension(self):
        """Returns the extention this operator uses.

        Returns:
            str: Name of the extension this solver uses.

        """

        return self.extension

    def getSolverArgs(self):
        """Returns the args array defined by the KL Operator.

        Returns:
            RTValArray: Args array defined by the KL Operator.

        """

        return self.args

    def getInputType(self, name):
        """Returns the type of input with the specified name."""
        for arg in self.args:
            if arg.connectionType.getSimpleType() == "In" and arg.name.getSimpleType() == name:
                return arg.dataType.getSimpleType()

        raise Exception("Could not find input argument %s in kl operator %s" % (name, self.getName()))

    def getOutputType(self, name):
        """Returns the type of output with the specified name."""
        for arg in self.args:
            if arg.connectionType.getSimpleType() == "Out" and arg.name.getSimpleType() == name:
                return arg.dataType.getSimpleType()

        raise Exception("Could not find output argument %s in kl operator %s" % (name, self.getName()))

    def getDefaultValue(self, name, RTValDataType, mode="arg"):
        """Returns the default RTVal value for this argument
        Only print debug if setting default inputs.  Don't care about outputs, really

        Args:
            name (str): Name of the input to get.
            mode (str): "inputs" or "outputs"

        Returns:
            RTVal

        """

        def isFixedArrayType(string):
            return bool(re.search(r'\[\d', string))

        # If attribute has a default value
        if self.solverRTVal.defaultValues.has("Boolean", name).getSimpleType():

            RTVal = ks.convertFromRTVal(self.solverRTVal.defaultValues[name])

            if RTVal.isArray():
                # If RTValDataType is variable array, but default value is fixed array, convert it
                if isFixedArrayType(RTVal.getTypeName().getSimpleType()) and not isFixedArrayType(RTValDataType):
                    RTValArray = ks.rtVal(RTValDataType)
                    if len(RTVal):
                        RTValArray.resize(len(RTVal))

                    for i in range(len(RTVal)):
                        RTValArray[i] = RTVal[i]

                    RTVal = RTValArray
            else:
                # Not totally sure why we need to do this, but we get None from getSimpleType from the RTVal
                # when we run it on it's own and use the type that we query.  Gotta investigate this further...
                RTVal = ks.convertFromRTVal(self.solverRTVal.defaultValues[name], RTTypeName=RTValDataType)

            logger.debug("Using default value for %s.%s.%s(%s) --> %s" % (self.solverTypeName, self.getName(), mode, name, RTVal))

            return RTVal

        defaultValue = ks.rtVal(RTValDataType)
        if name in self.inputs:  # Only warn for input types, OK that we generate default outputs, I think.  Maybe even safer.
            logger.warn("    Creating default value by generating new RTVal object of type: %s. You should set default values for %s.%s(%s) in your KL Operator." %
                (RTValDataType, self.solverTypeName, mode, name))

        return defaultValue

    def getInput(self, name):
        """Returns the input with the specified name.

        If there is no input value, it get the default RTVal and converts to
        python data

        Args:
            name (str): Name of the input to get.

        Returns:
            object: Input object.

        """

        if name in self.inputs and self.inputs[name] is not None:
            return self.inputs[name]

        def rt2Py(rtVal, rtType):

            if "[" in rtType:
                return []
            if rtType == "Xfo":
                return Xfo(rtVal)
            if rtType == "Mat44":
                return Mat44(rtVal)
            if rtType == "Vec2":
                return Vec2(rtVal)
            if rtType == "Vec3":
                return Vec3(rtVal)
            else:
                return rtVal.getSimpleType()

            #raise ValueError("Cannot convert rtval %s from %s" (rtVal, rtType))

        argDataType = None
        for arg in self.args:
            if arg.name.getSimpleType() == name:
                argDataType = arg.dataType.getSimpleType()
                break
        if argDataType is None:
            raise Exception("Cannot find arg %s for object %s" (arg, self.getName()))

        defaultVal = self.getDefaultValue(name, argDataType, mode="arg")
        pyVal = rt2Py(defaultVal, argDataType)

        return pyVal

    def generateSourceCode(self):
        """Returns the source code for a stub operator that will invoke the KL operator

        Returns:
            str: The source code for the stub operator.

        """

        # Start constructing the source code.
        opSourceCode = "dfgEntry {\n"

        # In SpliceMaya, output arrays are not resized by the system prior to
        # calling into Splice, so we explicily resize the arrays in the
        # generated operator stub code.
        for i in xrange(len(self.args)):
            arg = self.args[i]
            argName = arg.name.getSimpleType()
            argDataType = arg.dataType.getSimpleType()
            argConnectionType = arg.connectionType.getSimpleType()

            if argDataType.endswith('[]') and argConnectionType == 'Out':
                arraySize = len(self.getOutput(argName))
                opSourceCode += "  " + argName + ".resize(" + str(arraySize) + \
                    ");\n"

        opSourceCode += "  if(solver == null)\n"
        opSourceCode += "    solver = " + self.solverTypeName + "();\n"
        opSourceCode += "  solver.solve(\n"
        for i in xrange(len(self.args)):
            argName = self.args[i].name.getSimpleType()
            if i == len(self.args) - 1:
                opSourceCode += "    " + argName + "\n"
            else:
                opSourceCode += "    " + argName + ",\n"

        opSourceCode += "  );\n"
        opSourceCode += "}\n"

        return opSourceCode

    def evaluate(self):
        """Invokes the KL operator causing the output values to be computed.

        Returns:
            bool: True if successful.

        """

        # logger.debug("\nEvaluating kl operator [%s] of type [%s] from extension [%s]..." % (self.getName(), self.solverTypeName, self.extension))

        super(KLOperator, self).evaluate()

        def getRTVal(obj, asInput=True):
            if isinstance(obj, Object3D):
                if asInput:
                    return obj.globalXfo.getRTVal().toMat44('Mat44')
                else:
                    return obj.xfo.getRTVal().toMat44('Mat44')
            elif isinstance(obj, Xfo):
                return obj.getRTVal().toMat44('Mat44')
            elif isinstance(obj, MathObject):
                return obj.getRTVal()
            elif isinstance(obj, Attribute):
                return obj.getRTVal()
            elif type(obj) is bool:
                return ks.rtVal('Boolean', obj)
            elif type(obj) is int:
                return ks.rtVal('Integer', obj)
            elif type(obj) is float:
                return ks.rtVal('Scalar', obj)
            elif type(obj) is str:
                return ks.rtVal('String', obj)
            else:
                return obj #

        def validateArg(rtVal, argName, argDataType):
            """Validate argument types when passing built in Python types.

            Args:
                rtVal (RTVal): rtValue object.
                argName (str): Name of the argument being validated.
                argDataType (str): Type of the argument being validated.

            """

            # Validate types when passing a built in Python type
            if type(rtVal) in (bool, str, int, float):
                if argDataType in ('Scalar', 'Float32', 'UInt32', 'Integer'):
                    if type(rtVal) not in (float, int):
                        raise TypeError(self.getName() + ".evaluate(): Invalid Argument Value: " + str(rtVal) + " (" + type(rtVal).__name__ + "), for Argument: " + argName + " (" + argDataType + ")")

                elif argDataType == 'Boolean':
                    if type(rtVal) != bool:
                        raise TypeError(self.getName() + ".evaluate(): Invalid Argument Value: " + str(rtVal) + " (" + type(rtVal).__name__ + "), for Argument: " + argName + " (" + argDataType + ")")

                elif argDataType == 'String':
                    if type(rtVal) != str:
                        raise TypeError(self.getName() + ".evaluate(): Invalid Argument Value: " + str(rtVal) + " (" + type(rtVal).__name__ + "), for Argument: " + argName + " (" + argDataType + ")")

        argVals = []
        debug = []
        for i in xrange(len(self.args)):
            arg = self.args[i]
            argName = arg.name.getSimpleType()
            argDataType = arg.dataType.getSimpleType()
            argConnectionType = arg.connectionType.getSimpleType()

            if argDataType == 'EvalContext':
                argVals.append(ks.constructRTVal(argDataType))
                continue
            if argName == 'time':
                argVals.append(ks.constructRTVal(argDataType))
                continue
            if argName == 'frame':
                argVals.append(ks.constructRTVal(argDataType))
                continue

            if argConnectionType == 'In':
                if str(argDataType).endswith('[]'):
                    if argName in self.inputs and self.inputs[argName] is not None:
                        rtValArray = ks.rtVal(argDataType)
                        rtValArray.resize(len(self.inputs[argName]))
                        for j in xrange(len(self.inputs[argName])):
                            if self.inputs[argName][j] is None:
                                continue
                            rtVal = getRTVal(self.inputs[argName][j])

                            validateArg(rtVal, argName, argDataType[:-2])

                            rtValArray[j] = rtVal
                    else:
                        rtValArray = self.getDefaultValue(argName, argDataType, mode="arg")

                    argVals.append(rtValArray)
                else:
                    if argName in self.inputs and self.inputs[argName] is not None:
                        rtVal = getRTVal(self.inputs[argName])
                    else:
                        rtVal = self.getDefaultValue(argName, argDataType, mode="arg")

                    validateArg(rtVal, argName, argDataType)
                    argVals.append(rtVal)

            elif argConnectionType in ('IO', 'Out'):
                if str(argDataType).endswith('[]'):
                    if argName in self.outputs and self.outputs[argName] is not None:
                        rtValArray = ks.rtVal(argDataType)
                        rtValArray.resize(len(self.outputs[argName]))
                        for j in xrange(len(self.outputs[argName])):
                            if self.outputs[argName][j] is None:
                                continue
                            rtVal = getRTVal(self.outputs[argName][j], asInput=False)

                            validateArg(rtVal, argName, argDataType[:-2])

                            rtValArray[j] = rtVal
                    else:
                        rtValArray = self.getDefaultValue(argName, argDataType, mode="output")

                    argVals.append(rtValArray)
                else:
                    if argName in self.outputs and self.outputs[argName] is not None:
                        rtVal = getRTVal(self.outputs[argName], asInput=False)
                    else:
                        rtVal = self.getDefaultValue(argName, argDataType, mode="output")

                    validateArg(rtVal, argName, argDataType)

                    argVals.append(rtVal)
            else:
                raise Exception("Operator:'" + self.getName() + " has an invalid 'argConnectionType': " + argConnectionType)

            debug.append(
                {
                    argName: [
                        {
                            "dataType": argDataType,
                            "connectionType": argConnectionType
                        },
                        argVals[-1]
                    ]
                })

        try:
            # argstr = [str(arg) for arg in argVals]
            # logger.debug("%s.solve('', %s)" % (self.solverTypeName, ", ".join(argstr)))
            self.solverRTVal.solve('', *argVals)
        except Exception as e:

            errorMsg = "\nPossible problem with KL operator [%s]. Arguments:\n" % self.getName()
            errorMsg += pprint.pformat(debug, indent=4, width=800)
            logger.error(errorMsg)
            raise e

        # Now put the computed values out to the connected output objects.
        def setRTVal(obj, rtval):

            if isinstance(obj, Object3D):
                obj.xfo.setFromMat44(Mat44(rtval))
            elif isinstance(obj, Xfo):
                obj.setFromMat44(Mat44(rtval))
            elif isinstance(obj, Mat44):
                obj.setFromMat44(rtval)
            elif isinstance(obj, Attribute):
                if ks.isRTVal(rtval):
                    obj.setValue(rtval.getSimpleType())
                else:
                    obj.setValue(rtval)
            else:
                if hasattr(obj, '__iter__'):
                    logger.warning("Warning: Trying to set a KL port with an array directly.")

                logger.warning("Not setting rtval: %s\n\tfor output object: %s\n\tof KL object: %s\n." % \
                    (rtval, obj.getName(), self.getName()))

        for i in xrange(len(argVals)):
            arg = self.args[i]
            argName = arg.name.getSimpleType()
            argDataType = arg.dataType.getSimpleType()
            argConnectionType = arg.connectionType.getSimpleType()

            if argConnectionType != 'In':
                if argName in self.outputs and self.outputs[argName] is not None:
                    if str(argDataType).endswith('[]'):
                        for j in xrange(len(argVals[i])):
                            if len(self.outputs[argName]) > j and self.outputs[argName][j] is not None:
                                setRTVal(self.outputs[argName][j], argVals[i][j])
                    else:
                        setRTVal(self.outputs[argName], argVals[i])

        return True
