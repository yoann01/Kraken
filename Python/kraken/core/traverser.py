"""Kraken - traverser module.

Classes:
Traverser - Base Traverser.

"""

from kraken.core.objects.scene_item import SceneItem
from kraken.core.objects.object_3d import Object3D
from kraken.core.objects.components.component import Component
from kraken.core.objects.attributes.attribute_group import AttributeGroup
from kraken.core.objects.constraints.constraint import Constraint
from kraken.core.objects.operators.operator import Operator


class Traverser(object):
    """Kraken base traverser for any scene item.

    The traverser iterates through all root items and determines the
    dependencies between objects, building an ordered list as it goes. This
    order is then used by the builder to ensure that objects are created and
    evaluated in the correct order. Offset's will then be reliable.

    """

    def __init__(self, name='Traverser'):
        self._rootItems = []
        self.reset()

    # ==================
    # Property Methods
    # ==================
    @property
    def rootItems(self):
        """Gets the root items of this Traverser.

        Returns:
            list: The root items of this Traverser.

        """

        return self._xfo

    def addRootItem(self, item):
        """Adds a new root object to this Traverser

        Args:
            item (SceneItem): The SceneItem to add as a root item.

        Returns:
            bool: True if successful.

        """

        for rootItem in self._rootItems:
            if rootItem.getId() == item.getId():
                return False

        self._rootItems.append(item)

        return True

    def addRootItems(self, items):
        """Adds a bunch of root items to this Traverser

        Args:
            items (SceneItem[]): The SceneItems to add as root items.

        Returns:
            bool: True if successful.

        """

        for item in items:
            self.addRootItem(item)

        return True

    @property
    def items(self):
        """Gets the traversed items of this Traverser.

        Returns:
            list: The traversed items of this Traverser.

        """

        return self._items

    def getItemsOfType(self, typeNames):
        """Gets only the traversed items of a given type.

        Args:
            typeName (str): The name of the type to look for.

        Returns:
            list: The items in the total items list matching the given type.

        """

        if not isinstance(typeNames, list):
            typeNames = [typeNames]

        result = []
        for item in self._items:
            if item.isOfAnyType(typeNames):
                result.append(item)

        return result

    # =============
    # Traverse Methods
    # =============
    def reset(self):
        """Resets all internal structures of this Traverser."""

        self._visited = {}
        self._items = []

    def traverse(self, itemCallback=None, discoverCallback=None,
                 discoveredItemsFirst=True):
        """Visits all objects within this Traverser based on the root items.

        Args:
            itemCallback (func): A callback function to be invoked for each
                item visited.
            discoverCallback (func): A callback to return an array of children
                for each item.

        """

        self.reset()

        if discoverCallback is None:
            discoverCallback = self.discoverBySource

        for item in self._rootItems:
            self.__visitItem(item,
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)

        return self.items

    def __collectVisitedItem(self, item, itemCallback):
        """Doc String.

        Args:
            item (Type): information.
            itemCallback (Type): information.

        """

        if itemCallback is not None:
            itemCallback(item=item, traverser=self)

        self._items.append(item)

    def __visitItem(self, item, itemCallback, discoverCallback, discoveredItemsFirst):
        """Doc String.

        Args:
            item (Type): information.
            itemCallback (Type): information.
            discoverCallback (Type): information.
            discoveredItemsFirst (Type): information.

        Returns:
            Type: information.

        """

        if self._visited.get(item.getId(), False):
            return False

        self._visited[item.getId()] = True

        if hasattr(item, 'getParent') and item.getParent():
            # If this is an attribute and we have not traversed its parent AttributeGroup then skip this
            # and visit the parent so we get this attribute and all others from there (for the sake of attr order)
            if item.isTypeOf("Attribute") and not self._visited.get(item.getParent().getId(), False):
                self._visited[item.getId()] = False
                self.__visitItem(item.getParent(),
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)
                return False

            self.__visitItem(item.getParent(),
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)

        sourcedByConstraintOrOperator = False
        if discoveredItemsFirst:
            for source in item.getSources():
                if isinstance(source, (Constraint, Operator)):
                    sourcedByConstraintOrOperator = True
                    break

        if not discoveredItemsFirst or sourcedByConstraintOrOperator:
            self.__collectVisitedItem(item, itemCallback)

        if discoverCallback:
            if isinstance(item, AttributeGroup):
                discoveredItems = self.discoverChildren(item)
            else:
                discoveredItems = discoverCallback(item)

            if discoveredItems:
                for discoveredItem in discoveredItems:
                    self.__visitItem(discoveredItem,
                                     itemCallback,
                                     discoverCallback,
                                     discoveredItemsFirst)

        if discoveredItemsFirst and not sourcedByConstraintOrOperator:
            self.__collectVisitedItem(item, itemCallback)

        return True

    def discoverChildren(self, item):
        """Doc String.

        Args:
            item (Type): information.

        Returns:
            Type: information.

        """

        result = []

        if isinstance(item, Component):
            subItems = item.getItems().values()
            for subItem in subItems:
                if isinstance(subItem, AttributeGroup):
                    continue
                result.append(subItem)

        elif isinstance(item, Object3D):
            for i in xrange(item.getNumAttributeGroups()):
                result.append(item.getAttributeGroupByIndex(i))
            result += item.getChildren()

        if isinstance(item, AttributeGroup):
            for i in xrange(item.getNumAttributes()):
                result.append(item.getAttributeByIndex(i))

        return result

    def discoverBySource(self, item):
        """Doc String.

        Args:
            item (Type): information.

        Returns:
            Type: information.

        """

        result = []

        for source in item.getSources():

            if isinstance(source, Operator):
                for outputName in source.getOutputNames():
                    operatorOutputs = source.getOutput(outputName)
                    if not isinstance(operatorOutputs, list):
                        operatorOutputs = [operatorOutputs]
                    for operatorOutput in operatorOutputs:
                        if not isinstance(operatorOutput, SceneItem):
                            continue
                        result.append(operatorOutput)

            result.append(source)

        return result
