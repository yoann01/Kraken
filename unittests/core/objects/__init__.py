
import unittest

from attributes import suite as attributeSuite
from components import suite as componentSuite
from constraints import suite as constraintSuite
from operators import suite as operatorSuite

from test_component_group import suite as componentGroupSuite
from test_container import suite as containerSuite
from test_control import suite as controlSuite
from test_ctrlSpace import suite as ctrlSpaceSuite
from test_curve import suite as curveSuite
from test_hierarchy_group import suite as hierarchyGroupSuite
from test_joint import suite as jointSuite
from test_layer import suite as layerSuite
from test_locator import suite as locatorSuite
from test_object_3d import suite as object3DSuite
from test_rig import suite as rigSuite
from test_scene_item import suite as sceneItemSuite
from test_transform import suite as transformSuite

loadAttributeSuite = attributeSuite()
loadComponentSuite = componentSuite()
loadConstraintSuite = constraintSuite()
loadOperatorSuite = operatorSuite()

loadComponentGroupSuite = componentGroupSuite()
loadContainerSuite = containerSuite()
loadControlSuite = controlSuite()
loadCtrlSpaceSuite = ctrlSpaceSuite()
loadCurveSuite = curveSuite()
loadHierarchyGroupSuite = hierarchyGroupSuite()
loadJointSuite = jointSuite()
loadLayerSuite = layerSuite()
loadLocatorSuite = locatorSuite()
loadObject3DSuite = object3DSuite()
loadRigSuite = rigSuite()
loadSceneItemSuite = sceneItemSuite()
loadTransformSuite = transformSuite()


def suite():
    suites = [
        loadAttributeSuite,
        loadComponentSuite,
        loadConstraintSuite,
        loadOperatorSuite,
        loadContainerSuite,
        loadControlSuite,
        loadCtrlSpaceSuite,
        loadCurveSuite,
        loadHierarchyGroupSuite,
        loadJointSuite,
        loadLayerSuite,
        loadLocatorSuite,
        loadObject3DSuite,
        loadRigSuite,
        loadSceneItemSuite,
        loadTransformSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
