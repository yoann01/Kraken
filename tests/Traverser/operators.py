from kraken import plugins
from kraken.core.objects.locator import Locator
from kraken.core.objects.operators.kl_operator import KLOperator
from kraken.core.traverser.traverser import Traverser

locInA = Locator("locatorInA")
locInB = Locator("locatorInB")
locOutA = Locator("locatorOutA")
locOutB = Locator("locatorOutB")

operator = KLOperator("IK", "MultiPoseConstraintSolver", "Kraken")
operator.resizeInput('constrainers', 2)
operator.resizeOutput('constrainees', 2)
operator.setInput("constrainers", locInA, 0)
operator.setInput("constrainers", locInB, 1)
operator.setOutput("constrainees", locOutA, 0)
operator.setOutput("constrainees", locOutB, 1)

trav = Traverser()
trav.addRootItem(locOutA)
trav.addRootItem(locOutB)

def callback(**args):
    item = args.get('item', None)
    print 'Visited '+item.getDecoratedPath()

trav.traverse(itemCallback = callback)
