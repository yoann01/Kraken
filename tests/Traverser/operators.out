Failed to find DCC client. Falling back to Python client.
Visited locatorInA
Visited locatorInB
Visited locatorOutA
Visited locatorOutB
Visited IK
